include/swift/ABI/Task.h

```diff
@@ -148,19 +150,93 @@ class AsyncTask : public HeapObject, public Job {
+      /// In other words, even if it had a parent, the parent (and its parents)
+      /// are known to not contain any any more task locals, and thus any further
+      /// search beyond this task.
+      IsEmptyTerminal   = 0b00,
+      /// The storage pointer points at the next TaskLocalChainItem in this task.
+      IsNextItemPointer = 0b01,
+      /// The storage pointer points at a parent AsyncTask,
+      /// in which we should continue the lookup.
+      IsParentPointer   = 0b11
+  };
+
+  struct TaskLocalChainItem {
+      /// Mask used for the low status bits in a task local chain item.
+      static const uintptr_t statusMask = 0x03;
+
+      /// Metadata* to the Key of the task local value
+      uintptr_t keyType;
+      /// OpaqueValue* to the Value of the task local value
+      HeapObject *storage;
+
+      // TaskLocalChainItem*
+//      uintptr_t next;
+  };
+
+  /// Reserved for a chain of task local values,
+  /// if empty, the search should continue in this task's parent (if it has one).
+  TaskLocalChainItem *TaskLocal;
+
   AsyncTask(const HeapMetadata *metadata, JobFlags flags,
             TaskContinuationFunction *run,
             AsyncContext *initialContext)
     : HeapObject(metadata), Job(flags, run),
       ResumeContext(initialContext),
-      Status(ActiveTaskStatus()) {
+      Status(ActiveTaskStatus()),
+      TaskLocal(nullptr) {
     assert(flags.isAsyncTask());
   }
 
   void run(ExecutorRef currentExecutor) {
     ResumeTask(this, currentExecutor, ResumeContext);
   }
 
+  HeapObject* localGet(const Metadata *keyType) {
+    assert(keyType && "Task.Local key must not be null.");
+
+    if (!TaskLocal) {
+      return nullptr;
+    }
+
+    fprintf(stderr, "error: %s [%s:%d] TaskLocal=%d keyType=%d, lookup keyType=%d\n",
+            __FUNCTION__, __FILE_NAME__, __LINE__,
+            TaskLocal, TaskLocal->keyType, reinterpret_cast<uintptr_t>(keyType));
+    // TODO: can the contained list/dictionary for the key
+    if (TaskLocal->keyType == reinterpret_cast<uintptr_t>(keyType)) {
+      fprintf(stderr, "error: %s [%s:%d] return value TaskLocal=%d value=%d value.retainCount=%d\n",
+              __FUNCTION__, __FILE_NAME__, __LINE__,
+              TaskLocal, TaskLocal->storage, swift_retainCount(TaskLocal->storage));
+      return TaskLocal->storage;
+    }
+
+    if (hasChildFragment()) {
+      return childFragment()->getParent()->localGet(keyType);
+    } else {
+      // none of our parent tasks contained the value either, return null
+      return nullptr;
+    }
+  }
+
+  void localSet(const Metadata *keyType, HeapObject *value) {
+    assert(value && "Task local value must not be nil");
+
+    // TODO: proper implementation, with a linked list
+//    assert(!TaskLocal && "TaskLocal was already initialized");
+    TaskLocal = new TaskLocalChainItem;
+    TaskLocal->keyType = reinterpret_cast<uintptr_t>(keyType);
+    TaskLocal->storage = value;
```

Comment:
Got it, so: `valueType->vw_initializeWithCopy(head->storage, value);` however this crashes on me...

the way I'm passing through the value is:

```swift
    void append(const Metadata *keyType,
                /* +1 */ OpaqueValue *value, const Metadata *valueType) {
```

and I get this via

```c
SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
void swift_task_localValueAdd(AsyncTask* task,
                         const Metadata *keyType,
                         /* +1 */ OpaqueValue *value, const Metadata *valueType);
```

```swift
@_silgen_name("swift_task_localValueAdd")
public func _taskLocalValueAdd/*<Key>*/(
  _ task: Builtin.NativeObject,
  keyType: Any.Type/*Key.Type*/,
  value: __owned Any/*Key.Value*/,
  valueType: Any.Type/*Key.Value*/
) // where Key: TaskLocalKey
```

It is all Any's since I'm trying to support setting many values in one go, so they end up just Any; but for the runtime this should not really matter as far as I see... (passing an explicit type gives the same).

Or must I specifically use some different calling convention here? I did not understand that hint sadly too well :~

Reply:
The Metadata seems correct... so I must be messing up the `value` I'm passing (since I do __owned now...?);

The call has the correct metadata:

```
* thread #2, queue = 'com.apple.root.default-qos', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000010075e93e libswiftCore.dylib`initializeWithCopy for String at <compiler-generated>:0
    frame #1: 0x000000010016fc19 libswift_Concurrency.dylib`swift::OpaqueValue* swift::TargetMetadata<swift::InProcess>::vw_initializeWithCopy<swift::OpaqueValue*&, swift::OpaqueValue*&>(this=0x00000001009997f8, args=0x0000000101f06548, args=0x0000700006b079e8) const at ValueWitness.def:129:1
    frame #2: 0x000000010016e25d libswift_Concurrency.dylib`swift::AsyncTask::TaskLocalValuesFragment::append(this=0x0000000101f085f0, keyType=0x0000000100010468, value=0x0000700006b07c70, valueType=0x00000001009997f8) at Task.h:282:18
    frame #3: 0x000000010016e133 libswift_Concurrency.dylib`swift_task_localValueAdd(task=0x0000000101f08590, keyType=0x0000000100010468, value=0x0000700006b07c70, valueType=0x00000001009997f8) at Task.cpp:525:32
    frame #4: 0x0000000100160f4f libswift_Concurrency.dylib`$s12_Concurrency017_runChildTaskWithD6Locals8bindings9operationBoAA0D17LocalScopeBuilderV_xyYKctYlF(operation=<unavailable>) at Task.swift:453:5
```